features/storage/kvstore/kv_map/KVMap.cpp

```diff
@@ -82,21 +82,52 @@ void KVMap::deinit_partition(kv_map_entry_t *partition)
         return;
     }
 
+    if (partition->kv_config->kvstore_main_instance != NULL) {
+        partition->kv_config->kvstore_main_instance->deinit();
+        delete partition->kv_config->kvstore_main_instance;
+        partition->kv_config->kvstore_main_instance = NULL;
+    }
+
     if (partition->kv_config->external_store != NULL) {
         partition->kv_config->external_store->deinit();
+        delete partition->kv_config->external_store;
+        partition->kv_config->external_store = NULL;
+    }
+
+    if (partition->kv_config->internal_store != NULL) {
+        partition->kv_config->internal_store->deinit();
+        delete partition->kv_config->internal_store;
+        partition->kv_config->internal_store = NULL;
+    }
+
+    if (partition->kv_config->internal_bd != NULL) {
+        partition->kv_config->internal_bd->deinit();
+        delete partition->kv_config->internal_bd;
+        partition->kv_config->internal_bd = NULL;
     }
 
-    // TODO: this should be removed after FS APIs are standardized
     if (partition->kv_config->external_fs != NULL) {
         partition->kv_config->external_fs->unmount();
+        delete partition->kv_config->external_fs;
+        partition->kv_config->external_fs = NULL;
     }
 
-    if (partition->kv_config->internal_store != NULL) {
-        partition->kv_config->internal_store->deinit();
+    if (partition->kv_config->external_flashsim_bd != NULL) {
+        partition->kv_config->external_flashsim_bd->deinit();
+        delete partition->kv_config->external_flashsim_bd;
+        partition->kv_config->external_flashsim_bd = NULL;
     }
 
-    if (partition->kv_config->kvstore_main_instance != NULL) {
-        partition->kv_config->kvstore_main_instance->deinit();
+    if (partition->kv_config->external_slicing_bd != NULL) {
+        partition->kv_config->external_slicing_bd->deinit();
+        delete partition->kv_config->external_slicing_bd;
+        partition->kv_config->external_slicing_bd = NULL;
+    }
+
+    if (partition->kv_config->external_base_bd != NULL) {
+        partition->kv_config->external_base_bd->deinit();
```

Comment:
This is a bit dangerous:
1. In case of TDBStore, all underlying block devices are deinitialized, so we have a redundant deinit. It will work with all current block devices, as they have init ref counts, but may produce erroneous results with ones whose init/deinit scheme is not known to us.
2. In case of FileSystemStore, then its deinit does not unmount the underlying file system. However, this code deinitializes the underlying BD, leaving a mounted file system with a deinitialized BD, which is even worse of course. Now, at some point, we'll probably change the mount/unmount APIs in all file systems to an init/deinit scheme (there's a hanging PR for it), so FileSystemStore will behave as TDBStore, leaving us with what I said in (1).